packages/@ourworldindata/grapher/src/schema/migrations/helpers.ts

```diff
@@ -0,0 +1,58 @@
+import {
+    latestSchemaVersion,
+    outdatedSchemaVersions,
+} from "../defaultGrapherConfig"
+
+const allSchemaVersions = [...outdatedSchemaVersions, latestSchemaVersion]
+
+type LatestSchemaVersion = typeof latestSchemaVersion
+type OutdatedSchemaVersion = (typeof outdatedSchemaVersions)[number]
+type SchemaVersion = OutdatedSchemaVersion | LatestSchemaVersion
+
+type Schema =
+    `https://files.ourworldindata.org/schemas/grapher-schema.${SchemaVersion}.json`
+
+// we can't type configs that don't adhere to the latest schema as we don't know what they look like
+export type AnyConfig = Record<string, any>
+export type AnyConfigWithValidSchema = AnyConfig & {
+    $schema: Schema
+}
+
+const schemaVersionRegex =
+    /https:\/\/files\.ourworldindata\.org\/schemas\/grapher-schema\.(?<version>\d{3})\.json/m
+
+const isValidSchemaVersion = (version: string): version is SchemaVersion =>
+    allSchemaVersions.includes(version as any)
+
+export function getSchemaVersion(
+    config: AnyConfigWithValidSchema
+): SchemaVersion
+export function getSchemaVersion(config: AnyConfig): SchemaVersion | null
```

Comment:
Oh this is very neat - I didn't know typescript could do function overloading to narrow down types. Thanks for this!